### DISCUSSION TITLE
Corax search engine and highlightings

### DISCUSSION THREAD

**Discussion by marcuslindblom** [ravendb/ravendb#17640] [Q&A]
Is highlighting supported with Corax? I have highlightings on my search page but when I change search engine type to Corax I don't get any fragments anymore?

**Comment by arekpalinski** [comment]
Yes, it's supported. We had some issues around highlighting in Corax and it was fixed in https://github.com/ravendb/ravendb/pull/17571. It will be released in 6.0.1. If possible could you please check your scenario on 6.0 nightly? If the issue is still happening can you please share the details (index definition, query, sample docs etc) or write a failing test?

**Reply by marcuslindblom** [reply]
It seems to work, or at least I get some fragments. The fragments are not so user friendly because they can start in the middle of a sentence and they are concatenated so the text on my page is unreadable and does not make any sense. This is the result in Swedish, 150 max length and one fragment. The word `vaccinrna` does not exist in my original text at all so it puts together words that does not exist. Maybe this is a language thing?

<img width="745" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/d15713ed-4bb4-4691-8ae4-1a4c6c39055a">

What languages is supported out of the box. I build a search page that indexes mostly Swedish content.

**Reply by redknightlois** [reply]
@marcuslindblom It could be possible that the current Lucene implementation is using some token knowledge about the language that Corax is not. Can you provide us a small sample database or test case that showcase the issue?

**Reply by marcuslindblom** [reply]
@redknightlois No, I don't think that's the case. When using Lucene the fragments are also weird.
`vaccingheten` is not in my text and it's not a real word in Swedish.

<img width="614" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/149043b0-07e3-4341-bcef-fda071909e36">

**Reply by marcuslindblom** [reply]
Also, this example in english is also weird.
<img width="736" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/47f091f6-60ed-4ebe-839f-610f254e2d2c">
`Loopsn` is not in the text, but Loops is.

**Reply by redknightlois** [reply]
Can you provide me the documents and the index definition for reproducing this?

**Reply by marcuslindblom** [reply]
@redknightlois Sure, here is an example in English using Lucene.

```json
{
    "labels": [],
    "origin": {
        "id": "7a6c0383-cbc3-4e49-b535-922933e156dc",
        "collection": "roots"
    },
    "createdAt": "2023-11-07T08:21:10.2660000Z",
    "createdBy": "ffcb6a79-f1f3-4e21-b332-2713f31eeb81",
    "content": "<p>The Public Health Agency recommends that everyone who is 65 years and older, as well as staff in special accommodation for the elderly (SÄBO) and within home healthcare and home services, be offered a booster dose of the COVID-19 vaccine. It is the regions that decide when the different groups will be offered the vaccine according to the recommendation of the Public Health Agency.</p><p>The protection against severe illness and death from COVID-19 remains strong in all vaccinated groups. However, the accumulating knowledge increasingly shows that the vaccine's effectiveness decreases over time, but also that the development varies in different groups.</p><p>- Sweden vaccinated the most vulnerable first, and it is important that they continuously have good protection. We have already recommended a booster dose to the oldest group. Considering the increasing age-related risk and the time that has passed since many in the 65-79 age group and SÄBO, home healthcare, and home service personnel were vaccinated, it is time to recommend that these groups also receive a booster dose,\" says Anders Tegnell, State Epidemiologist and Head of Department at the Public Health Agency.</p><p>More groups are included</p><p>The risk of severe COVID-19 increases with age. The Public Health Agency has previously recommended a booster dose for individuals aged 80 and older, individuals receiving home healthcare or home services, and all residents in special accommodations for the elderly (SÄBO).</p><p>The Public Health Agency has decided to also recommend a third vaccine dose as a booster for individuals in the following groups:</p><p>- Individuals aged 65 to 79</p><p>- Staff in SÄBO and within home healthcare and home services</p><p>It is important to recommend a booster dose to staff in SÄBO and within home healthcare and home services to protect the care recipients within these facilities.</p><p>The recommendation is to offer an mRNA vaccine if at least six months have passed since the last vaccination. This applies regardless of the type of vaccine an individual has received previously and means that a person who previously received two doses of AstraZeneca's Vaxzevria vaccine is now recommended to receive Pfizer/Biontech's Comirnaty or Moderna's Spikevax. It has been previously recommended that individuals receiving a booster dose of the COVID-19 vaccine be administered half a dose when it comes to the mRNA vaccine Spikevax.</p><p>The regions are responsible for planning the vaccination efforts based on local conditions. If practical, it can be coordinated with seasonal influenza vaccinations.</p><p>Monitoring the vaccine's effectiveness</p><p>The overall goal of the vaccinations is to prevent severe illness and death, and this has been the guiding principle in prioritization. The Public Health Agency monitors the effectiveness of vaccinations in different groups and will make decisions regarding booster doses for individuals in more groups later on.</p><p>The basis for the Public Health Agency's decisions will be published next week.</p>",
    "heading": "A booster dose of the COVID-19 vaccine is recommended for more groups.",
    "publishedDate": "2023-11-07T08:21:55.1025800Z",
    "deleted": false,
    "displayName": "A booster dose of the COVID-19 vaccine is recommended for more groups.",
    "slug": "a-booster-dose-of-the-covid-19-vaccine-is-recommended-for-more-groups",
    "changedAt": "2023-11-07T08:36:07.7729240Z",
    "changedBy": "ffcb6a79-f1f3-4e21-b332-2713f31eeb81",
    "archived": false,
    "@metadata": {
        "@collection": "Articles",
        "Raven-Clr-Type": "Article, blog"
    }
}
```
The index definition looks like this.
```javascript
map('Articles', c => {
  let collection = c['@metadata']['@collection']
  let template = load('templates/' + collection, 'templates')
  if (!c.archived && template && !template.disableURL) {
    let id = Id(c)
    let visited = {}
    let slugs = []
    let origin = c.origin.id
    let displayName = c.displayName
    let publishedDate = c.publishedDate
    const terms = [];
    if (!c.search_settings || c.search_settings.show_page_in_site_search) {
        for (const field in template.searchFields) {
            terms.push(c[template.searchFields[field]].replace(/<[^>]*?>/gm, ' '));
        }
    }
    do {
      if (c.slug) {
          slugs.unshift(c.slug)
      }
      if (visited[c.origin.id])
        break
      visited[c.origin.id] = true
      c = load(c.origin.id, c.origin.collection)
    } while(c);
    return { id: id, displayName: displayName, url: slugs.shift() + slugs.join('/'), origin: origin, collection: collection, publishedDate: publishedDate, query: terms };
  }
});
```

This is the query I'm using
```c#
HighlightingOptions tagsToUse = new HighlightingOptions
{
    PreTags = new[] { @"<em>" },
    PostTags = new[] { "</em>" }
};
results = await session
  .Query<Result>("Content/Search")
  .Search(x => x.Query, query)
  .Highlight(x => x.Query,
    150, // Max length of each text fragment
    1, // Max number of fragments to return per document
    tagsToUse, // Highlighting options
    out Highlightings managerHighlights)
  .ProjectInto<Result.Hit>()
  .ToListAsync();
```
Searching for `vaccine` results in a hit like this.

```html
<p class="mt-5 text-sm leading-6 text-gray-600">
dose of the COVID-19 <em>vaccine</em>hat decide when the different groups will be offered the <em>vaccine</em> according to the</p>
```
<img width="614" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/d2080f0b-123f-44fc-ae79-8504e3f1399f">

One of the problems I see with this is that `vaccine` is highlighted but it is joined together with `hat`. With these settings I think a readable sentence that is understandable for the user is something that I expect from a search engine.

**Reply by arekpalinski** [reply]
Do you set the field options for `Query` field in the index definition?

https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/highlighting#highlight-results---map-index

**Reply by marcuslindblom** [reply]
This is the settings I have. Now this is for english with Lucene so the search highlighting does not seem to work there as well.
<img width="1882" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/f5043c4c-65ce-4840-991f-33bf5653e612">
And this is from the studio
<img width="1646" alt="image" src="https://github.com/ravendb/ravendb/assets/319720/9f0c9a78-c737-493a-bf55-f13b621f9e87">
As you can see it does not work well in the studio either.

**Reply by arekpalinski** [reply]
@marcuslindblom

The result that you got appears to be the concatenation of two fragments:

- "be offered a booster dose of the COVID-19 vaccine."
- "It is the regions that decide when the different groups will be offered the vaccine"

which resulted in: 
dose of the COVID-19 <b style="background:yellow">vaccine</b>hat decide when the different groups will be offered the <b style="background:yellow">vaccine</b> 

Note that "vaccinehat" origins from: "COVID-19 vaccine" and "that decide when" (without 't'). We got it from internals of Lucene:

![image](https://github.com/ravendb/ravendb/assets/376551/c7c83d61-6b24-4c11-8e06-e3f27d7a77ea)

It looks that the issue is related to the max length of the fragment - you specify `150`. Once I changed the query to use `250`:

```
from index 'Content/Search' where search(query, 'vaccine') include highlight(query, 250, 1)
```

then the result is more expected I think:

![image](https://github.com/ravendb/ravendb/assets/376551/26b4e5f9-3a61-47e8-8779-88d57969519f)

So it looks to be a matter of the highlighting parameters that we provide to Lucene

Note: I used simplified version of your index:

```
map('Articles', c => {
    return { query: c.content };
  });
```

**Reply by ayende** [reply]
I think that the problem is that we are scanning the text directly, and we aren't breaking on word boundary

**Reply by marcuslindblom** [reply]
@arekpalinski I have done some experiments with the fragment length but I can't say that I constantly get good results. If you add ten articles with different content you will see that the fragments is unreadable from time to time even with longer fragment length.

I think the problem exist in both Lucene and Corax.

**Reply by marcuslindblom** [reply]
@arekpalinski are you planning to fix this?

**Reply by arekpalinski** [reply]
It's an issue with longer fragments apparently. We'd need to investigate how exactly it's handled inside Lucene. We might want to look into that in the future. Given that it involves the internals of Lucene so we'd need to take into account the backward compatibility. 

Please note it's not any kind of regression in 6.0 version - it's been like this forever. You mentioned that your app has the usage of the highlighting already, right? You didn't spot this issue before or something got changed meanwhile on your side?

**Reply by marcuslindblom** [reply]
Yes, it seems like it's never worked like it should. Did you copy this feature to Corax because it does not work there either?

**Reply by arekpalinski** [reply]
Corax has its own implementation of this feature. It might be easier for us to fix it in Corax.

**Reply by loraderon** [reply]
> Corax has its own implementation of this feature. It might be easier for us to fix it in Corax.

I think it would be worth to only fix this (and future issues) in Corax and leave this and other features "as is" in Lucene. Then you do not need to focus so much on breaking changes (since Corax is new and people can switch to Lucene if required).

**Reply by marcuslindblom** [reply]
I agree. But as it works now it is unusable in both Corax and Lucene.

**Reply by arekpalinski** [reply]
I'm not sure if I called it unusable. The highlighted terms seems to be valid, right? The issue is with the fragments around the highlights if the fragments are longer / not long enough (but even with longer fragments you didn't get satisfactory results).

The highlighting results are returned additionally to the query results. As long as the query results are correct they are usable I believe. Maybe this is why nobody reported the issue with the fragments so far. Anyway, I opened the issue for checking this:

https://issues.hibernatingrhinos.com/issue/RavenDB-21695/Highlighting-Fragments-around-highlighted-terms-arent-always-readable
